google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1beta2/StreamWriterV2.java

```diff
@@ -171,8 +196,29 @@ public void run() {
                         "Stream is closed due to " + connectionFinalStatus.toString())));
         return requestWrapper.appendResult;
       }
+
+      ++this.inflightRequests;
+      this.inflightBytes += requestWrapper.messageSize;
       waitingRequestQueue.addLast(requestWrapper);
       hasMessageInWaitingQueue.signal();
+
+      // Maybe block until we are below inflight limit.
+      while (this.inflightRequests >= this.maxInflightRequests
+          || this.inflightBytes >= this.maxInflightBytes) {
+        try {
+          inflightReduced.await(100, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+          log.warning(
+              "Interrupted while waiting for inflight quota. Stream: "
+                  + streamName
+                  + " Error: "
+                  + e.toString());
+          throw new StatusRuntimeException(
+              Status.fromCode(Code.CANCELLED)
+                  .withCause(e)
+                  .withDescription("Interrupted while waiting for quota."));
+        }
+      }
```

Comment:
Try to make the coder a little bit cleaner, let me try:

var flight_request/bytes;
{
  lock;
  addToQueue();
  inflight_requests/bytes = xxx;
}

// This should be normal cases (without extra lock operation).
if (inflight_request/bytes < threshold) {
  return;
}

// Too busy case
waitConditionWithATimeout(xxx)

Reply:
Done.